habitat-lab/habitat/tasks/rearrange/actions/grip_actions.py

```diff
@@ -158,3 +161,178 @@ def _grasp(self):
+
+        # Get the camera transformation
+        cam_T = self.get_camera_transform()
+        # Get object location in camera frame
+        cam_obj_pos = cam_T.inverted().transform_point(obj_pos).normalized()
+
+        # Get angle between (normalized) location and the vector that the camera should
+        # look at
+        obj_angle = self.angle_between(cam_obj_pos, self.center_cone_vector)
+
+        return obj_angle
+
+    def get_camera_transform(self):
+        if isinstance(self._sim.robot, SpotRobot):
+            cam_info = self._sim.robot.params.cameras[
+                "articulated_agent_arm_depth"
+            ]
+        elif isinstance(self._sim.robot, StretchRobot):
+            cam_info = self._sim.robot.params.cameras["robot_head"]
+        else:
+            raise NotImplementedError(
+                "This robot does not have GazeGraspAction."
+            )
+
+        # Get the camera's attached link
+        link_trans = self._sim.robot.sim_obj.get_link_scene_node(
+            cam_info.attached_link_id
+        ).transformation
+        # Get the camera offset transformation
+        offset_trans = mn.Matrix4.translation(cam_info.cam_offset_pos)
+        cam_trans = link_trans @ offset_trans @ cam_info.relative_transform
+
+        return cam_trans
+
+    def determine_center_object(self):
+        cam_pos = None
+        if self._distance_from == 'camera':
+            """Determine if an object is at the center of the frame and in range"""
+            if isinstance(self._sim.robot, SpotRobot):
+                cam_pos = (
+                    self._sim.agents[0]
+                    .get_state()
+                    .sensor_states["articulated_agent_arm_rgb"]
+                    .position
+                )
+            elif isinstance(self._sim.robot, StretchRobot):
+                cam_pos = (
+                    self._sim.agents[0]
+                    .get_state()
+                    .sensor_states["robot_head_depth"]
+                    .position
+                )
+            else:
+                raise NotImplementedError(
+                    "This robot does not have GazeGraspAction."
+                )
+
+        if isinstance(self._sim.robot, StretchRobot):
+            panoptic_img = self._sim._sensor_suite.get_observations(
+                self._sim.get_sensor_observations()
+            )['robot_head_panoptic']
+        else:
+            raise NotImplementedError(
+                "This robot dose not have GazeGraspAction."
+        )
+
+        height, width = panoptic_img.shape[:2]
+
+        if self._center_square_width == 1:
+            center_obj_id = panoptic_img[height // 2,  width // 2] - self._instance_ids_start
+        else:
+            # check if any pixel within the center square has a valid pixel
+            if isinstance(self._sim.robot, StretchRobot):
+                obj_seg = self._task.sensor_suite.get_observations(
```

Comment:
We should see if there's a better way of doing this, won't this force the sim to re-render all observations? The panoptic sensor should probably be enough here since it has all instance IDs.

Reply:
Updated, now rendering sim observations only once with this. Let me also see if there's a better way

Reply:
I think we should cache the observations in `RearrangeTask` as `prev_observations` or something like that and then just use that here to avoid recomputation.